test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeSystemArraycopyTests.java

```diff
@@ -0,0 +1,396 @@
+			ifVtArraySrc[i] = new SomePrimitiveValueClass(i*8);
+
+			ifArray1[i] = new SomeIdentityClass(i*9);
+			ifArray2[i] = new SomePrimitiveValueClass(i*10);
+		}
+	}
+
+	static private void checkResults(Object[] arr1, Object[] arr2) {
+		for (int i=0; i<arr1.length; ++i) {
+			assertEquals(arr1[i], arr2[i]);
+		}
+	}
+
+	static private void testVTVT(SomePrimitiveValueClass[] src, SomePrimitiveValueClass[] dst) {
+		System.arraycopy(src, 0, dst, 0, ARRAY_SIZE);
+	}
+
+	static private void testVTIF(SomePrimitiveValueClass[] src, SomeInterface[] dst) {
+		System.arraycopy(src, 0, dst, 0, ARRAY_SIZE);
+	}
+
+	static private void testIFVT(SomeInterface[] src, SomePrimitiveValueClass[] dst) {
+		System.arraycopy(src, 0, dst, 0, ARRAY_SIZE);
+	}
+
+	static private void testIDID(SomeIdentityClass[] src, SomeIdentityClass[] dst) {
+		System.arraycopy(src, 0, dst, 0, ARRAY_SIZE);
+	}
+
+	static private void testIDIF(SomeIdentityClass[] src, SomeInterface[] dst) {
+		System.arraycopy(src, 0, dst, 0, ARRAY_SIZE);
+	}
+
+	static private void testIFID(SomeInterface[] src, SomeIdentityClass[] dst) {
+		System.arraycopy(src, 0, dst, 0, ARRAY_SIZE);
+	}
+
+	static private void testIFIF(SomeInterface[] src, SomeInterface[] dst) {
+		System.arraycopy(src, 0, dst, 0, ARRAY_SIZE);
+	}
+
+	static private void testIDOBJ(SomeIdentityClass[] src) {
+
+		Object[] dst = new Object[ARRAY_SIZE];
+
+		for (int i=0; i < ARRAY_SIZE; i++) {
+			dst[i] = new SomeIdentityClass(i*11);
+		}
+
+		System.arraycopy(src, 0, dst, 0, ARRAY_SIZE);
+
+		checkResults(src, dst);
+	}
+
+	static private void testOBJID(SomeIdentityClass[] dst) {
+
+		Object[] src = new Object[ARRAY_SIZE];
+
+		for (int i=0; i < ARRAY_SIZE; i++) {
+			src[i] = new SomeIdentityClass(i*11);
+		}
+
+		System.arraycopy(src, 0, dst, 0, ARRAY_SIZE);
+
+		checkResults(src, dst);
+	}
+
+	static private void testVTOBJ(SomePrimitiveValueClass[] src) {
+
+		Object[] dst = new Object[ARRAY_SIZE];
+
+		for (int i=0; i < ARRAY_SIZE; i++) {
+			dst[i] = new SomePrimitiveValueClass(i*11);
+		}
+
+		System.arraycopy(src, 0, dst, 0, ARRAY_SIZE);
+
+		checkResults(src, dst);
+	}
+
+	static private void testOBJVT(SomePrimitiveValueClass[] dst) {
+
+		Object[] src = new Object[ARRAY_SIZE];
+
+		for (int i=0; i < ARRAY_SIZE; i++) {
+			src[i] = new SomePrimitiveValueClass(i*11);
+		}
+
+		System.arraycopy(src, 0, dst, 0, ARRAY_SIZE);
+
+		checkResults(src, dst);
+	}
+
+	static private void testIFOBJ(SomeInterface[] src) {
+
+		Object[] dst = new Object[ARRAY_SIZE];
+
+		for (int i=0; i < ARRAY_SIZE; i++) {
+			dst[i] = new SomePrimitiveValueClass(i*11);
+		}
+
+		System.arraycopy(src, 0, dst, 0, ARRAY_SIZE);
+
+		checkResults(src, dst);
+	}
+
+	static private void testOBJIF(SomeInterface[] dst) {
+
+		Object[] src = new Object[ARRAY_SIZE];
+
+		for (int i=0; i < ARRAY_SIZE; i++) {
+			src[i] = new SomePrimitiveValueClass(i*11);
+		}
+
+		System.arraycopy(src, 0, dst, 0, ARRAY_SIZE);
+
+		checkResults(src, dst);
+	}
```

Comment:
I think it would be good to have cases where the Object array comes from a parameter, so that it's not known at compile-time whether the component type really is java/lang/Object.